### PR TITLE
Correct purpose differentiation between Icarus and Shelley

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

adp-766

# Overview

- [x] correcting account -> hardened derivation
- [x] differentiate between Icarus and Shelley 

# Comments

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs [hunk 1]
```diff
@@ -389,13 +391,17 @@ addresses mkAddress =
     . indexedKeys
   where
     toDerivationPath ix = NE.fromList $ map DerivationIndex
-        [ getIndex purposeBIP44
-        , getIndex purposeCIP1852
-        , 0   -- corresponds to account number
+        [ getIndex $ getPurpose @k
+        , getIndex coinTypeAda
+        , getIndex $ minBound @(Index 'Hardened 'AccountK)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:396]: Awe. Indeed. Sorry I didn't catch that earlier :facepalm:

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs [hunk 2]
```diff
@@ -389,13 +391,17 @@ addresses mkAddress =
...
         , fromIntegral $ fromEnum $ role @c
         , getIndex ix
         ]
 
+-- It is used for geting purpose for a given key.
+class GetPurpose key where
+    getPurpose :: Index 'Hardened 'PurposeK
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:403]: :+1:

**Review by KtorZ — APPROVED** [review]

**Comment by KtorZ** [comment]
Note: approving now, but our test coverage should've been / should be sufficient to cover this (thus, please make sure to upgrade the integration or unit test suite to cover this, they should fail before this change and pass after :pray:)

**Comment by piotr-iohk** [comment]
I have added additional assertions to check if derivation path is valid for Shelley / Icarus and Byron addresses in the integration tests.

**Review by piotr-iohk — APPROVED** [review]
lgtm.

**Pull request merged**